SimPEG/base/pde_simulation.py

```diff
@@ -0,0 +1,529 @@
+                M_prop_deriv = self.mesh.get_edge_inner_product_deriv(
+                    np.ones(self.mesh.n_cells)
+                )(np.ones(self.mesh.n_edges)) * getattr(self, f"{arg.lower()}Deriv")
+                setattr(self, stash_name, M_prop_deriv)
+            return __inner_mat_mul_op(
+                getattr(self, stash_name), u, v=v, adjoint=adjoint
+            )
+
+        setattr(cls, f"Me{arg}Deriv", MeDeriv_prop)
+
+        def MccIDeriv_prop(self, u, v=None, adjoint=False):
+            """
+            Derivative of `MccPropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mcc{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mcc{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mcc{arg}IDeriv", MccIDeriv_prop)
+
+        def MnIDeriv_prop(self, u, v=None, adjoint=False):
+            """
+            Derivative of `MnPropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mn{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mn{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mn{arg}IDeriv", MnIDeriv_prop)
+
+        def MfIDeriv_prop(self, u, v=None, adjoint=False):
+            """I
+            Derivative of `MfPropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mf{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mf{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mf{arg}IDeriv", MfIDeriv_prop)
+
+        def MeIDeriv_prop(self, u, v=None, adjoint=False):
+            """
+            Derivative of `MePropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Me{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Me{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Me{arg}IDeriv", MeIDeriv_prop)
+
+        @property
+        def _clear_on_prop_update(self):
```

Comment:
this will clear all of them on every model update? We might want to look and see if the mappings are set before choosing what needs to be cleared (e.g. in an EM simulation / inversion, we don't need to be clearing the mu matrices, unless you are actually inverting for magnetic permeability)

Reply:
This part isn't what is actually called when the property is updated, it just represents a list of items that should be cleared if you update the physical property. This functions as the `_clear_on_mu_update` and `_clear_on_sigma_update` properties on the current `BaseEMSimulation` class.

Down in the `BaseElectricalPDESimulation` you'll see where it performs the deletions, which executes in the same way as our current simulations.

Reply:
:+1: Thanks for clarifying!